<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="UTF-8" />
    <title>Title</title>
    <style type="text/css"></style>
  </head>
  <body>
    <script type="text/javascript">
      /* 
    then/catch 都是正常返回 resolved，里边有报错则返回 rejected
    then/catch 是根据 resolved/rejected来进行触发，resolved触发 then，rejected触发 catch
    如果then中有报错 会返回 rejected
    如果catch中没有报错 会返回 resolved
    
 */

      // const p1 = Promise.resolve().then(()=>{
      //   return 100
      // })
      // p1.then(()=>{
      //   console.log(100);
      // })

      // const p2 = Promise.resolve().then(()=>{
      //   throw new Error('err100')
      // })
      // p2.then(()=>{
      //   console.log(456);
      // }).catch((err)=>{
      //   console.log(err, 'err');
      // })
      // const p3 = Promise.reject('my error').catch((err)=>{
      //   console.log(err);  //此时返回 resolved，如果报错了 则返回reject
      // })

      function foo(i) {
        return new Promise((resolve, reject) => {
          setTimeout(() => {
            resolve(i);
          }, 1000);
        });
      }

      (async function () {
        // 异步遍历

        for (var i = 0; i < 3; i++) {
          var i;
          const res = foo(i);
          console.log("ok", i);
        }
      })();
      // async function async1() {
      //   console.log("async1 start"); // 2
      //   await async2(); //await这行是同步代码，执行后函数内 await下边代码将放入微任务队列
      //   // 关键在这一步，它相当于微任务放在 micro task queue 微任务队列 中最后执行
      //   console.log("async1 end");  // 微任务 1  ， 6
      // }

      // async function async2() {
      //   console.log("async2"); // 3
      // }

      // console.log("script start");  // 1

      // setTimeout(function () {
      //   console.log("setTimeout");  //宏任务 1  8
      // }, 0);

      // async1();

      // new Promise(function (resolve) {
      //   console.log("promise1"); // 4
      //   resolve();
      //   console.log("？？？"); // 4
      // }).then(() => {
      //   console.log("promise2"); // 微任务 2 ， 7  微任务完毕 开始 DOM渲染后开始 Evnent Loop开始宏任务
      // });
      // console.log("script end"); // 5   这步结束后 同步代码全部完毕 开始执行微任务

      //   ---------- 题目 ----------
      // 第一题
      Promise.resolve()
        .then(() => {
          console.log(1);
        })
        .catch(() => {
          console.log(2);
        })
        .then(() => {
          console.log(3);
        });

      // 第二题
      Promise.resolve()
        .then(() => {
          console.log(1);
          throw new Error("erro1");
        })
        .catch(() => {
          console.log(2);
        })
        .then(() => {
          console.log(3);
        });

      // 第三题
      Promise.resolve()
        .then(() => {
          console.log(1);
          throw new Error("erro1");
        })
        .catch(() => {
          console.log(2);
        })
        .catch(() => {
          // 注意这里是 catch
          console.log(3);
        });
    </script>
  </body>
</html>
